src/hopen/syntax/mustache.cljc

```diff
@@ -0,0 +1,132 @@
+(ns hopen.syntax.mustache
+  #?(:clj (:import [java.util.regex Pattern]))
+  (:require [clojure.string :as str]))
+
+;;------------------------------------------------------------------------------
+;; General utilities
+;;------------------------------------------------------------------------------
+
+;; TODO!!! Provide a js implementation
+(defn re-seq-pos
+  "Returns a lazy sequence of regex matches like `re-seq`, but also returns the offsets of
+  the groups/captures."
+  [pattern string]
+  (let [m (re-matcher pattern string)]
+    ((fn step []
+       (when (. m find)
+         (cons {:start (.start m)
+                :end (.end m)
+                :groups (into [] (for [g (range (inc (.groupCount m)))]
+                                   (.group m g)))}
+               (lazy-seq (step))))))))
+
+(defn conj-last
+  "Conj `x` into the last element of the `coll`.
+  `coll` should be a seq of seqs."
+  [coll x]
+  (conj (pop coll) (conj (peek coll) x)))
+
+
+;;------------------------------------------------------------------------------
+;; Namespace specific functions
+;;------------------------------------------------------------------------------
+
+;; TODO!!! Provide a js implementation
+(defn- compile-regex [start-delim end-delim]
+  (Pattern/compile
+   (str
+    (Pattern/quote start-delim)
+    "(.*)"
+    (Pattern/quote end-delim))))
+
+(defn- parse-delim-change-tag
+  "Try to parse the tag-text as a delimiter change command.
+  Returns '(start-delim end-delim) or nil."
+  [tag-text]
+  (let [result (re-matches #"=(.*)\s+(.*)=" tag-text)]
+    (when (= 3 (count result))
+      (drop 1 result))))
+
+(defn- retrieve-all-tags
+  "Retrieve a list of all tags and their offsets.
+  Any delimiter changes will be dealt with here.
+
+  A tag is represented by a single entry returned by `re-seq-pos`,
+  which looks like:
+  {:start 'character offset into text where the tag starts'
+   :end 'character offset into text where the tag ends'
+
+   :groups ['full text of tag with delimiters' 'text of tag without delimiters']}"
+  [text]
+
+  ;; A tag is represented by a single entry returned by `re-seq-pos`.
+  (let [update-tags-with-offset (fn [offset tags]
+                                  (if (zero? offset)
+                                    tags
+                                    (map #(-> %
+                                              (update :start + offset)
+                                              (update :end + offset))
+                                         tags)))]
+    (loop [regex (compile-regex "{{" "}}")
+           offset 0
+           tags []]
+
+      (let [;; Given the tags we can find with our current delimiter.
+            ;; Collect all tags that do not denote delim change.
+            ;; When we do find a delim change, stop further processing.
+            possible-tags (->> (re-seq-pos regex (subs text offset))
```

Comment:
Just going to make a note here that `subs` is called, a significant portion of the template text will likely be copied (at least on the JVM). Would be nice to fix this at some point.

Reply:
I disagree. On the JVM, strings are immutable and Java takes advantage about that by avoiding duplication of the characters. The strings are wrapping the original char buffer with 2 indexes for the beginning and the end of the string w.r.t. the original char buffer.

Really, no problem there.

Reply:
https://dzone.com/articles/changes-stringsubstring-java-7

Apparently String.substring no longer return slices. Full copies are made instead as of java 7.

However, this is only an issue for “very large” inputs that frequently switches its delimiters. It is probably a non-issue. Just making a note of it here.

Reply:
> This all changed in May of 2012 or Java 7u6.

Wow .. I am soooo outdated